### PR TITLE
Added SingleBatchCursor

### PR DESCRIPTION
**Pull request by rozza**
Part of a wider Command Cursor refactoring

JAVA-5159

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java [hunk 1]
```diff
@@ -19,19 +19,26 @@
...
-        this.firstBatch = firstBatch;
-        isTrue("Empty Cursor", firstBatch.getCursor() == null);
+    static <R> AsyncSingleBatchCursor<R> createEmptyAsyncSingleBatchCursor(final int batchSize) {
+        return new AsyncSingleBatchCursor<>(emptyList(), batchSize);
+    }
+
+    private final List<T> batch;
+    private final int batchSize;
+
+    private volatile boolean hasNext = true;
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:36]: Valentin has commented previously that volatile variables like this are unnecessary.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java [hunk 2]
```diff
@@ -43,13 +50,12 @@ public void close() {
...
-            List<T> results = firstBatch.getResults();
-            firstBatch = null;
-            callback.onResult(results, null);
+        } else if (hasNext && !batch.isEmpty()) {
+            hasNext = false;
+            callback.onResult(batch, null);
         } else {
             closed = true;
-            callback.onResult(null, null);
+            callback.onResult(emptyList(), null);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:58]: Why the change from `null` to `emptyList()` here?  This make it inconsistent with  `AsyncQueryBatchCursor`.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java (hunk 1 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:36]: Previously queryresult was volatile and that has been replaced by hasNext.

Not sure where previously it was mentioned but usage wasn't changed in AsyncSingleBatchQueryCursor, so I kept the same behaviour

File: driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:58]: So the end state is emptyList for all?

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:58]: This will bring it inline with both sync and async cursors.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:58]: I peeked ahead at the last PR, and I see some null handling in the sync cursor, e.g.

```
this.nextBatch = commandCursor.getResults().isEmpty() ? null : commandCursor.getResults();
```

Are you looking to change that to allow `nextBatch` to just be set to the empty list if it happens to be empty? 

Relatedly, should we:

* Update the javadoc of `AsyncBatchCursor#next` to reflect this change (iirc you had done that in the previous PR)?
* replace the `results != null` conditional in `BatchCursorFlux#recurseCursor` and the `result == null` check in `BatchCursor#next` with assertions?

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java:58]: Added #402 with the javadoc change and recurseCursor change.

`nextBatch` in the sync cursor is a single variable representing a Tenary of values:
 - the nextBatch that can be returned to next
 - if the nextBatch has any results
 - if nextBatch has been processed
 
That said sync cursor `#next()` will never return a null, and now the async cursor won't either.

**Review by jyemin — APPROVED** [review]

**Comment by rozza** [comment]
Closed and rolled up to #401
